### PR TITLE
[AutoTune] Gpu bind parallel

### PR DESCRIPTION
**Pull request by zhhsplendid**
This PR did 3 things:

1. Modify MultiLevelTiling AutoGenRule to bind GPU block/threads. **Note**: this AutoGenRule found the `Reorder` schedule primitive may have bug, so it is still not fully completed. See the comments in `multi_level_tiling.cc`

2. Add `SearchState` so we could pass not only `ModuleExpr` data during auto tune search.

3. Set`CudaAxisInfo` to auto tune `LoweredFunc` to fix `test_auto_tuner` GPU target error.

Note: This PR doesn't add bind information if AutoGenRule didn't add it. We simply added a single thread GPU kernel. This should be improved after we have complete AutoGenRule(s) and check the auto-generated binding informations. We kept it as TODO after we optimize the CINN speed.

### CODE REVIEW

File: cinn/auto_schedule/auto_tuner_test.cc [hunk 1]
```diff
@@ -76,7 +80,9 @@ class TestAutoTuner : public ::testing::Test {
     compile_options.Apply(result);
     ASSERT_EQ(2, compile_options.groups.size());
     ASSERT_EQ(2, compile_options.lowered_funcs.size());
-
+    std::cout << "Print lowered_funcs before building" << std::endl;
+    std::cout << compile_options.lowered_funcs[0][0] << std::endl;
+    std::cout << compile_options.lowered_funcs[1][0] << std::endl;
```
Comment (CtfGo) [inline cinn/auto_schedule/auto_tuner_test.cc]: use vlog or remove them

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc [hunk 2]
```diff
@@ -127,33 +148,59 @@ ir::ModuleExpr MultiLevelTiling::Apply(int index) {
...
+  for (auto& t : tiles) {
+    std::reverse(t.begin(), t.end());
+    splited_loops.insert(splited_loops.end(), t.begin(), t.end());
+  }
 
-  ir_schedule_->Reorder(splited_loops);
-  VLOG(5) << "Finish Reorder in MultiLevelTiling";
+  // Reorder has some bugs in this case. Enable it after we fix it.
+  // ir_schedule_->Reorder(splited_loops);
+  // VLOG(5) << "Finish Reorder in MultiLevelTiling";
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: mark it as `TODO`

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc [hunk 3]
```diff
@@ -127,33 +148,59 @@ ir::ModuleExpr MultiLevelTiling::Apply(int index) {
...
+
+  int num_binds = std::min(bind_axis_.size(), tiles.size());
+  for (int i = 0; i < num_binds; ++i) {
+    //   Enable fuse and then bind after fixing reorder bug.
+    //   Expr fused = ir_schedule_->Fuse(tiles[i]);
+    //   ir_schedule_->Bind(fused, bind_axis_[i]);
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: as above

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc [hunk 4]
```diff
@@ -74,34 +74,8 @@ TEST(MultiLevelTile, SimpleLoops) {
...
-          }
-        }
-      }
-    }
-  }
-}
-)ROC";
-  EXPECT_EQ(utils::Trim(target_str), utils::Trim(expr_str));
+  std::string expr_str = ss.str();
+  std::cout << expr_str << std::endl;
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc]: why remove the comparison in result expr

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc [hunk 5]
```diff
@@ -145,47 +119,8 @@ TEST(MulitLevelTile, MatrixMultiply) {
...
-        }
-      }
-    }
-  }
-}
-)ROC";
-
-  EXPECT_EQ(utils::Trim(target_str), utils::Trim(expr_str));
+  std::string expr_str = ss.str();
+  std::cout << expr_str << std::endl;
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc]: as above

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h [hunk 6]
```diff
@@ -54,31 +61,61 @@ class MultiLevelTiling : public AutoGenRule {
...
+      T tmp   = pick[0];
+      pick[0] = pick[1];
+      pick[1] = tmp;
     }
     return pick;
   }
 
+  // Sample num_split integers whose product equals extent
+  template <typename T>
+  std::vector<T> SampleTileSplit(T extent, int num_split) const {
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h:86]: suggest to add tests for these two functions

File: cinn/auto_schedule/auto_tuner_test.cc (hunk 1 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/auto_tuner_test.cc]: Done. Use VLOG

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc (hunk 2 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: Done.

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc (hunk 3 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.cc]: Done

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc (hunk 4 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc]: As you have seen and commented above, the schedule primitives contain bugs and the generated Exprs would still change in the future. I still have to edit the result Expr if I keep the check. I would put the check after we totally finish search space.

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc (hunk 5 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc]: As you have seen from above, the schedule primitives contain bugs and the generated Exprs would still change in the future. I still have to edit the result Expr if I keep the check. I would put the check after we totally finish search space.

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h (hunk 6 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h:86]: Added, done.

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc (hunk 4 above)
Comment (CtfGo) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling_test.cc]: OK

**Review by CtfGo — DISMISSED** [review]
LGTM for the update in `auto_schedule`

**Review by CtfGo — APPROVED** [review]

**Pull request merged**